### PR TITLE
58 dockerizing frontend and backend

### LINKED ISSUE #58: Dockerizing Frontend and Backend
### Problem Description

Both the frontend (Next.js) and the backend (Python with FastAPI) applications are not yet containerized, hindering their deployment on AWS ECS and EKS.

### Solution

Develop Dockerfiles for both the frontend and backend applications, enabling them to be containerized and run in isolated environments.

### Rationale

Containerization facilitates easy deployment, scaling, and ensures consistency across different environments, essential for a microservices architecture on AWS.

#### Frontend:

- Create a Dockerfile for the Next.js application, including the build step and serving the build via a web server like Nginx.

#### Backend:

- Create a Dockerfile for the FastAPI application, ensuring it includes all dependencies and uses a suitable base image for Python applications.
- Push both Docker images to AWS ECR for storage and versioning.

---

### LINKED ISSUE #59: Dockerizing Frontend and Backend for Deployment
**Description**  
_As a software developer,_  
_so that I can facilitate easy deployment, scaling, and ensure consistency across different environments for our microservices architecture on AWS,_  
_I want to develop Dockerfiles for both the frontend (Next.js) and backend (Python with FastAPI) applications, enabling them to be containerized and run in isolated environments._

**Acceptance Criteria**
- [ ] A Dockerfile is created for the Next.js frontend application, including steps for building the application and serving it via a web server like Nginx.
- [ ] A Dockerfile is created for the FastAPI backend application, ensuring it includes all necessary dependencies and uses a suitable base image for Python applications.
- [ ] Both Docker images can be successfully built using their respective Dockerfiles, ensuring the applications operate correctly within the container environment.
- [ ] Both Docker images are tested to confirm they meet the operational requirements of their respective applications in various environments.
- [ ] Both Docker images are pushed to AWS Elastic Container Registry (ECR) for storage, versioning, and easy deployment to AWS ECS or EKS.
- [ ] Documentation is updated to include instructions for building the Docker images, pushing them to ECR, and deploying them to ECS or EKS.

### PR DESCRIPTION
**Pull request by Ivan-E-Johnson**
# Overview
This PR introduces Docker support for both the frontend and backend components of our project. By containerizing these elements, we enable their deployment on AWS ECS and EKS, facilitating easier deployment, scaling, and management.

# Implementation

Docker compose
Docker frontend 
Docker Backend

# Testing
Manual testing to ensure 

# Problems Faced
We ran into so many problems with networking and cors 

closes #58 
closes #59

### CODE REVIEW

**Review by mbrzus — APPROVED** [review]
good work!

**Comment by powersaudrey25** [comment]
I am still seeing a CORS error 
Access to fetch at 'http://localhost:8000/users/' (redirected from 'http://localhost:8000/users') from origin 'http://localhost:3000' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.

**Comment by powersaudrey25** [comment]
> I am still seeing a CORS error Access to fetch at 'http://localhost:8000/users/' (redirected from 'http://localhost:8000/users') from origin 'http://localhost:3000' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.

Changing 
    environment:
      - NEXT_PUBLIC_BACKEND_URL=http://backend:8000
     to
    environment:
      - NEXT_PUBLIC_BACKEND_URL=http://0.0.0.0:8000
fixed it.

Not sure if changing it back to backend was intended?

File: docker-compose.yml [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+      dockerfile: Dockerfile_Frontend
+    ports:
+      - "3000:3000"
+    networks:
+      - application-network
+    depends_on:
+      - backend
+      - database
+    environment:
+      - NEXT_PUBLIC_BACKEND_URL=http://backend:8000
```
Comment (powersaudrey25) [inline docker-compose.yml:33]: should this be 0.0.0.0:8000?

Comment (Ivan-E-Johnson) [reply docker-compose.yml:33]: no it was intentional to ensure that we are not exposing it to everyone using port 8000. did you try and rebuild all of your images? Are you available for a quick teams chat?

**Comment by powersaudrey25** [comment]
> > I am still seeing a CORS error Access to fetch at 'http://localhost:8000/users/' (redirected from 'http://localhost:8000/users') from origin 'http://localhost:3000' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.
> 
> Changing environment: - NEXT_PUBLIC_BACKEND_URL=http://backend:8000 to environment: - NEXT_PUBLIC_BACKEND_URL=http://0.0.0.0:8000 fixed it.
> 
> Not sure if changing it back to backend was intended?

just had to rebuild, disregard

**Review by powersaudrey25 — APPROVED** [review]
looks good! thanks for figuring this out!

**Pull request merged**